File: 295. Find Median from Data Stream.java
/**
 * Uses double priorityqueue for maintaining a min heap and a max heap
 * the median is either the top of maxheap or the median of minheap and maxheap
 * Time O(logn) for add, O(1) for getting the median
 * Space O(n)
 */
class MedianFinder {
    PriorityQueue<Integer> maxHeap, minHeap;

    /** initialize your data structure here. */
    public MedianFinder() {
        maxHeap = new PriorityQueue<>(Collections.reverseOrder());
        minHeap = new PriorityQueue<>();
    }
    
    public void addNum(int num) {
        if (maxHeap.isEmpty() || num <= maxHeap.peek()) {
            maxHeap.offer(num);
        } else {
            minHeap.offer(num);
        }
        if (maxHeap.size() - minHeap.size() > 1) {
            minHeap.offer(maxHeap.poll());
        }
        if (minHeap.size() > maxHeap.size()) {
            maxHeap.offer(minHeap.poll());
        }
    }
    
    public double findMedian() {
        if (maxHeap.size() == minHeap.size()) {
            return (double)(maxHeap.peek() + minHeap.peek()) / 2.0;
        } else {
            return (double)maxHeap.peek();
        }
    }
}

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder obj = new MedianFinder();
 * obj.addNum(num);
 * double param_2 = obj.findMedian();
 */

File: 295. Find Median from Data Stream.py
from heapq import heappush, heappop
class MedianFinder:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.count = 0
        self.higher = []
        self.lower = []

    def addNum(self, num: int) -> None:
        if self.count == 0:
            heappush(self.lower, -num)
            self.count+=1
        elif self.count == 1:
            if num > -1*self.lower[0]:
                heappush(self.higher, num)
                self.count+=1
            else:
                temp = -1*heappop(self.lower)
                heappush(self.higher, temp)
                heappush(self.lower, -num)
                self.count+=1
        elif self.count % 2 == 0:
            if num > self.higher[0]:
                temp = heappop(self.higher)
                heappush(self.higher, num)
                heappush(self.lower, -temp)
                self.count+=1
            else:
                heappush(self.lower, -num)
                self.count+=1
        else:
            if num < -1*self.lower[0]:
                temp = -1*heappop(self.lower)
                heappush(self.lower, -num)
                heappush(self.higher, temp)
                self.count+=1
            else:
                heappush(self.higher, num)
                self.count+=1
        

    def findMedian(self) -> float:
        if self.count % 2 == 0:
            return (-1*self.lower[0] + self.higher[0])/2
        else:
            return -1.0 * self.lower[0]


# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()